#Exercise 4: Feature Engineering using data aggregation on AMES Housing dataset
In this exercise, we will learn how to create 2 new features by aggregating the data from different columns.

The dataset we will be using in this exercise is the Ames Housing dataset and it can be found on our GitHub repository - https://github.com/PacktWorkshops/The-Data-Science-Workshop/blob/master/Chapter12/Dataset/ames_iowa_housing.csv

Note
The dataset used for this exercise is the Ames Housing dataset compiled by Dean De Cock: https://www.openintro.org/stat/data/?data=ames 

This dataset contains the list of residential home sales in the city of Ames, Iowa between 20016 and 2010.

More information about each variable can be found here:
http://jse.amstat.org/v19n3/decock/DataDocumentation.txt



1. Open on a new Colab notebook and import the pandas and altair packages

In [0]:
import pandas as pd

2. Assign the link to the dataset to a variable called 'file_url'




In [0]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter12/Dataset/ames_iowa_housing.csv'

3. Using the read_csv .method() from the package pandas, load the dataset into a new DataFrame called 'df'

In [0]:
df = pd.read_csv(file_url)

4. Perform data aggregation to find the maximum 'SalePrice' for each Neighborhood' and 'YrSold' using .groupby.agg() methods and save the results in a new DataFrame called df_agg.

In [0]:
df_agg = df.groupby(['Neighborhood', 'YrSold']).agg({'SalePrice': 'max'}).reset_index()

5. Rename the column of df_agg to 'Neighborhood', 'YrSold' and 'SalePriceMax'

In [0]:
df_agg.columns = ['Neighborhood', 'YrSold', 'SalePriceMax']

6. Print out the first 5 rows of df_agg

In [6]:
df_agg.head()

Neighborhood  YrSold  SalePriceMax
0      Blmngtn    2006        264561
1      Blmngtn    2007        194201
2      Blmngtn    2008        191000
3      Blmngtn    2009        192500
4      Blmngtn    2010        192000

7. Merge the original DataFrame df to df_agg using a left join (how='left') on the columns 'Neighborhood' and 'YrSold' using the merge() method and save the results into a new DataFrame called df_new

In [0]:
df_new = pd.merge(df, df_agg, how='left', on=['Neighborhood', 'YrSold'])

8. Print out the first 5 rows of df_new

In [8]:
df_new.head()

Id  MSSubClass MSZoning  ...  SaleCondition  SalePrice SalePriceMax
0   1          60       RL  ...         Normal     208500       287000
1   2          20       RL  ...         Normal     181500       294000
2   3          60       RL  ...         Normal     223500       287000
3   4          70       RL  ...        Abnorml     140000       250000
4   5          60       RL  ...         Normal     250000       350000

[5 rows x 82 columns]

9. Create a new column called 'SalePriceRatio' by dividing 'SalePrice' to 'SalePriceMax'

In [0]:
df_new['SalePriceRatio'] = df_new['SalePrice'] / df_new['SalePriceMax']

10. Print out the first 5 rows of df_new

In [10]:
df_new.head()

Id  MSSubClass MSZoning  ...  SalePrice  SalePriceMax SalePriceRatio
0   1          60       RL  ...     208500        287000       0.726481
1   2          20       RL  ...     181500        294000       0.617347
2   3          60       RL  ...     223500        287000       0.778746
3   4          70       RL  ...     140000        250000       0.560000
4   5          60       RL  ...     250000        350000       0.714286

[5 rows x 83 columns]

11. Perform data aggregation to find the maximum 'LotArea' for each Neighborhood' and 'YrSold' using .groupby.agg() methods and save the results in a new DataFrame called df_agg2.

In [0]:
df_agg2 = df.groupby(['Neighborhood', 'YrSold']).agg({'LotArea': 'max'}).reset_index()

12. Rename the column of df_agg2 to 'Neighborhood', 'YrSold' and 'LotAreaMax' and print out the first 5 columns

In [12]:
df_agg2.columns = ['Neighborhood', 'YrSold', 'LotAreaMax']
df_agg2.head()

Neighborhood  YrSold  LotAreaMax
0      Blmngtn    2006        4045
1      Blmngtn    2007        3922
2      Blmngtn    2008        3182
3      Blmngtn    2009        3684
4      Blmngtn    2010        3182

13. Merge the original DataFrame df to df_agg2 using a left join (how='left') on the columns 'Neighborhood' and 'YrSold' using the merge() method and save the results into a new DataFramce called df_final

In [0]:
df_final = pd.merge(df_new, df_agg2, how='left', on=['Neighborhood', 'YrSold'])

14. Create a new column called 'LotAreaRatio' by dividing 'LotArea' to 'LotAreaMax'

In [0]:
df_final['LotAreaRatio'] = df_final['LotArea'] / df_final['LotAreaMax']

15. Print out the first 5 rows of df_final for the following columns: Id, Neighborhood, YrSold, SalePrice, SalePriceMax, SalePriceRatio, LotArea, LotAreaMax, LotAreaRatio

In [15]:
df_final[['Id', 'Neighborhood', 'YrSold', 'SalePrice', 'SalePriceMax', 'SalePriceRatio', 'LotArea', 'LotAreaMax', 'LotAreaRatio']].head()

Id Neighborhood  YrSold  ...  LotArea  LotAreaMax  LotAreaRatio
0   1      CollgCr    2008  ...     8450       13125      0.643810
1   2      Veenker    2007  ...     9600       17542      0.547258
2   3      CollgCr    2008  ...    11250       13125      0.857143
3   4      Crawfor    2006  ...     9550       16560      0.576691
4   5      NoRidge    2008  ...    14260       14303      0.996994

[5 rows x 9 columns]

Great! We just created 2 new features that give the ratio of SalePrice and LotArea for a property compared to the highest one that was sold in the same year and same neighborhood. We can now easily compare them together. For instance from the output above we can notice the fifth property was almost as close as the biggest property sold (LotAreaRatio is 0.996994) but its sale price was significantly lower than the highest one (SalePriceRatio is 0.714286). This indicates that other features on the property have an impact on the sale price.